File: cv_engine/api/schemas/drafts.py
"""WorkingDrafts, validation, and approval over HTTP.

The draft document itself travels as an object, for the same reason the
analysis and the plan do in `schemas/analyses.py`: it is a versioned domain
document, and a second hand-written copy of its schema in the HTTP layer could
only drift from the one that is enforced.

Nothing here carries a filesystem path, and nothing here carries the ETag. The
token is a header, which is what makes it usable by a conditional request.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import Field, model_validator

from ...domain.models import ClaimStyle, ClaimType, OmissionReason, SelectionOutcome
from .health import HttpSchema


class GenerateWorkingDraftRequest(HttpSchema):
    """What `POST /applications/{id}/working-draft/generate` accepts.

    Both source IDs are explicit. A generate that resolved the latest analysis
    and plan for itself could build from a plan the user has never seen -
    §5.4's fifth acceptance item is exactly that commands take explicit source
    IDs.
    """

    job_analysis_id: str
    selection_plan_id: str
    provider: Literal["deterministic", "openai"] = "deterministic"


class RegenerateSectionRequest(HttpSchema):
    """What `POST /working-drafts/{id}/regenerate-section` accepts (§14).

    The draft's exact version and content hash are in the body rather than in
    `If-Match`, on the same reasoning as `validate` and `approve`: this is an
    action on a resource, not a conditional replacement of it, and an action
    that accepted `If-Match: *` would be the lost update the header exists to
    prevent.

    The analysis and plan are explicit. A regeneration that resolved them for
    itself could rewrite a section against a plan the user never saw.
    """

    application_id: str
    expected_edit_version: int
    expected_content_hash: str
    job_analysis_id: str
    selection_plan_id: str
    section: str = Field(max_length=200)
    instruction: str = Field(default="", max_length=2000)


class RegenerateClaimRequest(HttpSchema):
    """What `POST /working-drafts/{id}/regenerate-claim` accepts (§14)."""

    application_id: str
    expected_edit_version: int
    expected_content_hash: str
    job_analysis_id: str
    selection_plan_id: str
    claim_id: str = Field(max_length=200)
    instruction: str = Field(default="", max_length=2000)


class ClaimPatchRequest(HttpSchema):
    """One claim's replacement content.

    Free text with no fact behind it is accepted and saved as a pending claim
    carrying the reason it could not be authorized; it is never discarded.
    """

    claim_id: str
    fact_ids: list[str] = []
    text: str | None = None
    template_id: str | None = None
    template_version: str | None = None


class UpdateWorkingDraftRequest(HttpSchema):
    """The structured patch `PATCH /working-drafts/{id}` applies as one edit.

    Removals travel with the edits rather than in a command of their own.
    Product spec §10 makes removal one of the three resolutions for free text
    nothing could authorize, and §14 commits an autosave patch as a single edit
    against a single expected version - a second command would need its own
    token and could interleave with the save already in flight.

    Only one of the two lists has to be non-empty. Requiring both would make
    "remove this line" impossible to express without also rewriting one.
    """

    claim_edits: list[ClaimPatchRequest] = []
    claim_removals: list[str] = Field(
        default=[],
        description=(
            "Claims to delete outright. Only an unauthorized section claim may "
            "be removed this way; a claim the fact selection authorizes is a "
            "412 naming apply-selection-change, and the headline and contacts "
            "are structural."
        ),
    )

    @model_validator(mode="after")
    def validate_patch(self) -> UpdateWorkingDraftRequest:
        """Refused as a bad request, before any of it is applied.

        The command carries the same invariant for the CLI, but a contradiction
        that only surfaced there would reach the client as a 500 rather than as
        the `422` an unusable request deserves.
        """
        if not self.claim_edits and not self.claim_removals:
            raise ValueError("a patch must edit or remove at least one claim")
        both = {edit.claim_id for edit in self.claim_edits} & set(self.claim_removals)
        if both:
            raise ValueError(f"a patch cannot both edit and remove the same claim: {sorted(both)}")
        return self


class ApplySelectionChangeRequest(HttpSchema):
    """The user's explicit fact decisions for a deterministic re-selection.

    Two lists, not three, on the same reasoning as the analysis review form:
    `selected` is what the resulting plan reports, and explicit inclusion is a
    pin.
    """

    expected_edit_version: int
    pinned_fact_ids: list[str] = []
    excluded_fact_ids: list[str] = []


class WorkingDraftVersionRequest(HttpSchema):
    """The exact version a command is addressed to."""

    expected_edit_version: int


class ReplaceWorkingDraftRequest(WorkingDraftVersionRequest):
    """Replacement names its own compatible analysis and plan (§14).

    The route is `POST /applications/{id}/working-draft/replace`, so the path
    carries the Application and the body carries the draft. Both are explicit
    and neither is inferred from the other: the client states which draft of
    which Application it means, and a pair that does not match is a `412`.

    `keep_previous` is the Keep decision: it materializes the immutable
    historical snapshot before the replacement is attempted.
    """

    working_draft_id: str
    job_analysis_id: str
    selection_plan_id: str
    keep_previous: bool = False
    provider: Literal["deterministic", "openai"] = "deterministic"


class ApproveDraftRequest(WorkingDraftVersionRequest):
    """Approval names the ValidationRun whose result it is relying on.

    The run is not created here. Approval verifies that the run describes this
    exact draft version and content and that it passed; a run approval created
    for itself could only ever agree with approval.
    """

    validation_run_id: str


class DraftClaimResponse(HttpSchema):
    """One editable line: what a claim edit addresses, plus what it currently is.

    `claim_type` and `style` are the domain's closed sets rather than `str`, so
    a client's status labels stay exhaustive over them and a claim type added
    here fails that client's build instead of reaching a screen untranslated.
    """

    claim_id: str
    style: ClaimStyle
    text: str
    claim_type: ClaimType
    fact_ids: list[str]
    pending_reason: str | None = None


class DraftSectionResponse(HttpSchema):
    name: str
    claims: list[DraftClaimResponse]


class DraftOutlineResponse(HttpSchema):
    """The document's editable structure, derived from `source` on every read.

    Not a competing copy of the draft: it stores nothing, it is computed from
    the same object `source` carries, and it holds only what an edit can
    address. `source` remains the whole versioned document for anything that
    needs more than the editor does.
    """

    headline: DraftClaimResponse
    contacts: list[DraftClaimResponse]
    sections: list[DraftSectionResponse]


class DraftFactResponse(HttpSchema):
    """One fact this draft uses, or one its SelectionPlan considered.

    `text` is nullable: a fact the store can no longer resolve is already
    reported as a stale reason by the state projection, and a read that raised
    over it would turn an explainable staleness into a `500`.

    `outcome` is null for a fact no SelectionPlan ranked - a contact, or a fact
    a manual relink attached. That null is what says no include/exclude decision
    applies to it, so no second flag is needed to say the same thing.
    """

    fact_id: str
    text: str | None = None
    linked_claim_ids: list[str] = []
    section: str | None = None
    outcome: SelectionOutcome | None = None
    reason: OmissionReason | None = None


class WorkingDraftFactsResponse(HttpSchema):
    """§20 candidate accounting: the union of linked facts and plan candidates.

    Neither set covers the other. Contacts come from the candidate context and
    appear in no SelectionPlan; an omitted candidate appears in no claim. The
    editor needs both - one to say what backs a line, the other to say what
    could be added to one.
    """

    working_draft_id: str
    application_id: str
    selection_plan_id: str
    language: str
    facts: list[DraftFactResponse]


class WorkingDraftResponse(HttpSchema):
    id: str
    application_id: str
    job_analysis_id: str
    selection_plan_id: str
    parent_revision_id: str | None = None
    source: dict[str, Any]
    outline: DraftOutlineResponse
    edit_version: int
    content_hash: str
    active: bool
    created_at: str
    updated_at: str
    latest_validation_run_id: str | None = None
    latest_validation_passed: bool | None = None


class WorkingDraftUpdateResponse(HttpSchema):
    """The new token, and which claims were saved as pending."""

    application_id: str
    working_draft_id: str
    edit_version: int
    content_hash: str
    selection_plan_id: str
    pending_claim_ids: list[str] = []


class SelectionChangeResponse(HttpSchema):
    application_id: str
    working_draft_id: str
    edit_version: int
    content_hash: str
    selection_plan_id: str
    plan: dict[str, Any]


class ArchivedWorkingDraftResponse(HttpSchema):
    application_id: str
    working_draft_id: str
    edit_version: int
    content_hash: str
    artifact_version_id: str


class ValidationRunResponse(HttpSchema):
    """A ValidationRun as data. `passed=false` arrives as `200` (§22)."""

    application_id: str
    working_draft_id: str
    validation_run_id: str
    edit_version: int
    content_hash: str
    passed: bool
    report: dict[str, Any]


class ApprovalResponse(HttpSchema):
    application_id: str
    revision_id: str
    version: int
    markdown_artifact_version_id: str
    manifest_artifact_version_id: str
    decision_record_id: str

File: frontend/src/api/drafts.ts
import { queryOptions } from "@tanstack/react-query";

import { type ApiPath, apiRequest } from "./client";
import { type QueuedOperation, queuedOperation } from "./operations";
import type {
  ApplySelectionChangeRequest,
  ClaimPatch,
  DraftClaim,
  Operation,
  RegenerateClaimRequest,
  RegenerateSectionRequest,
  SelectionChange,
  WorkingDraft,
  WorkingDraftFacts,
  WorkingDraftUpdate,
} from "./contracts";

/* The draft and the token that authorizes writing to it, kept together.

   A read that returned only the body would leave the ETag to be captured somewhere else,
   and the one thing an optimistic save must never do is send a token that came from a
   different read than the content the user was looking at. */
export interface DraftRead {
  draft: WorkingDraft;
  etag: string | null;
}

export const workingDraftQueryKey = (workingDraftId: string) =>
  ["working-draft", workingDraftId] as const;

export const workingDraftFactsQueryKey = (workingDraftId: string) =>
  ["working-draft-facts", workingDraftId] as const;

const workingDraftPath = (workingDraftId: string): ApiPath =>
  `/api/v1/working-drafts/${encodeURIComponent(workingDraftId)}`;

export const workingDraftQueryOptions = (workingDraftId: string) =>
  queryOptions({
    queryKey: workingDraftQueryKey(workingDraftId),
    queryFn: async ({ signal }): Promise<DraftRead> => {
      const response = await apiRequest<WorkingDraft>(workingDraftPath(workingDraftId), { signal });
      return { draft: response.data, etag: response.etag };
    },
  });

/* §20 candidate accounting. Separate from the draft read because it answers a different
   question - what backs a line, and what could be added to one - and because the fact
   renderings are Knowledge, which moves on its own schedule rather than with an edit. */
export const workingDraftFactsQueryOptions = (workingDraftId: string) =>
  queryOptions({
    queryKey: workingDraftFactsQueryKey(workingDraftId),
    queryFn: async ({ signal }) => {
      const response = await apiRequest<WorkingDraftFacts>(
        `${workingDraftPath(workingDraftId)}/facts` as ApiPath,
        { signal },
      );
      return response.data;
    },
  });

/* The preview is framed, not fetched: the browser loads this URL inside a sandboxed
   iframe, so it never passes through `apiRequest` and never becomes a string this code
   holds. The version is in the query string only so that a save produces a different URL
   and the frame reloads; the server ignores it and answers for the current version. */
export const draftPreviewSrc = (workingDraftId: string, editVersion: number): string =>
  `${workingDraftPath(workingDraftId)}/preview?v=${editVersion}`;

/* Every claim in the document, in reading order, headline and contacts included - the
   same set `draft_claims` walks on the backend, so the editor and the projection cannot
   disagree about what a claim is. */
export const outlineClaims = (draft: WorkingDraft): DraftClaim[] => [
  draft.outline.headline,
  ...draft.outline.contacts,
  ...draft.outline.sections.flatMap((section) => section.claims),
];

export interface DraftPatch {
  claim_edits: ClaimPatch[];
  claim_removals: string[];
}

export interface DraftUpdate {
  update: WorkingDraftUpdate;
  etag: string | null;
}

/* §14 autosave. The ETag is the caller's, and the caller's obligation is that it came
   from the read whose content the user was editing - `If-Match: *` is refused by the
   server precisely because a save that matched anything is the lost update the header
   exists to prevent. */
export const updateWorkingDraft = async (
  workingDraftId: string,
  etag: string,
  patch: DraftPatch,
): Promise<DraftUpdate> => {
  const response = await apiRequest<WorkingDraftUpdate>(workingDraftPath(workingDraftId), {
    method: "PATCH",
    body: patch,
    etag,
  });

  return { update: response.data, etag: response.etag };
};

/* The overlay the next SelectionPlan is built from. Both lists are absolute, not deltas:
   `apply_selection_change` hands them to `create_selection_plan`, which plans from them
   alone, so a change that sent only what moved would silently drop every earlier
   decision. They are read back out of the accounting rather than remembered in component
   state, which is what keeps them true after a reload. */
export const selectionOverlay = (
  facts: WorkingDraftFacts,
): { pinned_fact_ids: string[]; excluded_fact_ids: string[] } => ({
  pinned_fact_ids: facts.facts
    .filter((fact) => fact.outcome === "pinned")
    .map((fact) => fact.fact_id),
  excluded_fact_ids: facts.facts
    .filter((fact) => fact.reason === "excluded_by_user")
    .map((fact) => fact.fact_id),
});

/* §14: a deterministic re-selection. `201` because the change creates an immutable
   SelectionPlan and rebuilds the draft from it in one write - there is no Operation to
   poll and no `Location` to verify.

   A draft carrying manual wording is a `412`: a deterministic rebuild would replace the
   user's own sentences with the engine's, which is the case §14 sends to regeneration. */
export const applySelectionChange = async (
  workingDraftId: string,
  expectedEditVersion: number,
  overlay: { pinned_fact_ids: string[]; excluded_fact_ids: string[] },
): Promise<SelectionChange> => {
  const body: ApplySelectionChangeRequest = {
    expected_edit_version: expectedEditVersion,
    ...overlay,
  };
  const response = await apiRequest<SelectionChange>(
    `${workingDraftPath(workingDraftId)}/apply-selection-change` as ApiPath,
    { method: "POST", body },
  );

  return response.data;
};

/* §14 regeneration: an AI Operation over one exact draft version.

   All three parts of the draft's identity are sent - ID, `edit_version`, and content hash
   - because that is what the Operation freezes. An autosave that lands while the provider
   is answering makes activation fail as `SOURCE_CHANGED` rather than overwriting the
   user's edit, and that is the behaviour these arguments buy.

   The analysis and the plan are explicit for the same reason `create_draft` names them: a
   regeneration that resolved them for itself could rewrite a section against a plan the
   user never saw. `provider` is omitted, as everywhere else, because the server owns that
   default. */
const regenerate = async (
  path: ApiPath,
  /* `instruction` is omitted rather than sent empty: the server defaults it, and
     restating a default here would be a second copy of a policy this client does not
     own. The `Omit` still binds every field name to the generated contract. */
  body: Omit<RegenerateSectionRequest, "instruction"> | Omit<RegenerateClaimRequest, "instruction">,
  idempotencyKey: string,
): Promise<QueuedOperation> =>
  queuedOperation(await apiRequest<Operation>(path, { method: "POST", body, idempotencyKey }));

export const regenerateSection = async (
  draft: WorkingDraft,
  section: string,
  idempotencyKey: string,
): Promise<QueuedOperation> =>
  regenerate(
    `${workingDraftPath(draft.id)}/regenerate-section` as ApiPath,
    {
      application_id: draft.application_id,
      expected_edit_version: draft.edit_version,
      expected_content_hash: draft.content_hash,
      job_analysis_id: draft.job_analysis_id,
      selection_plan_id: draft.selection_plan_id,
      section,
    },
    idempotencyKey,
  );

export const regenerateClaim = async (
  draft: WorkingDraft,
  claimId: string,
  idempotencyKey: string,
): Promise<QueuedOperation> =>
  regenerate(
    `${workingDraftPath(draft.id)}/regenerate-claim` as ApiPath,
    {
      application_id: draft.application_id,
      expected_edit_version: draft.edit_version,
      expected_content_hash: draft.content_hash,
      job_analysis_id: draft.job_analysis_id,
      selection_plan_id: draft.selection_plan_id,
      claim_id: claimId,
    },
    idempotencyKey,
  );
